File: Dynamic Programming/coin-combination-1/iterative_dp.py
import resource, sys
resource.setrlimit(resource.RLIMIT_STACK, (2**29,-1))

inf = 1e10
n, target = map(int, input().split())
coins = list(map(int, input().split()))

dp = [0] * (target+1) 
dp[0] = 1

for i in range(target+1):
	for coin in coins:
		if(i - coin >= 0):
			pick = dp[i - coin]
			dp[i] += pick 
print(dp[target])

File: Dynamic Programming/coin-combination-1/iterative_dp.cpp
#include <bits/stdc++.h>
using namespace std;

const long long inf = 1e10;
const long long modulo = 1e9 + 7;

int main() {
	long long n, target;
	cin >> n >> target;
	long long coin;
	vector<long long> coins;
	for(int i = 0; i < n; i++) {
		cin >> coin;
		coins.push_back(coin);
	}
	vector<int> dp(target+1, 0);
	dp[0] = 1;
	for(int i = 1; i < target+1; i++) {
		for(int coin: coins) {
			if((i - coin) >= 0) dp[i] = (dp[i] + dp[i - coin]) % modulo;
		}
	}
	cout << dp[target] << "\n";
} 